### PR TITLE
Bug 1984414: Log resource diffs on update only in reconcile mode

### PR DESCRIPTION
**Pull request by arjunrn**
Logging of resource diffs was introduced to detect [hotlooping in the CVO](https://github.com/openshift/cluster-version-operator/pull/561). But this increases the verbosity of logging during updates when the resources have changed which is not useful. The logging of the diffs is more useful in the reconcile mode. Also the diff is generated with [ObjectDiff](https://github.com/kubernetes/apimachinery/blob/235edae7dd90601011bbe3bcd6f84f7dc857b034/pkg/util/diff/diff.go#L57) which is deprecated. The `ObjectDiff` function now just uses [cmp.Diff](https://pkg.go.dev/github.com/google/go-cmp/cmp#Diff) internally and all instances of `ObjectDiff` have been replaced with `cmp.Diff`.

### CODE REVIEW

File: lib/resourceapply/apiext.go [hunk 1]
```diff
@@ -3,17 +3,17 @@ package resourceapply
...
 	apiextclientv1 "k8s.io/apiextensions-apiserver/pkg/client/clientset/clientset/typed/apiextensions/v1"
 	apierrors "k8s.io/apimachinery/pkg/api/errors"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
-	"k8s.io/apimachinery/pkg/util/diff"
 	"k8s.io/klog/v2"
 	"k8s.io/utils/pointer"
 )
 
-func ApplyCustomResourceDefinitionv1(ctx context.Context, client apiextclientv1.CustomResourceDefinitionsGetter, required *apiextv1.CustomResourceDefinition) (*apiextv1.CustomResourceDefinition, bool, error) {
+func ApplyCustomResourceDefinitionv1(ctx context.Context, client apiextclientv1.CustomResourceDefinitionsGetter, required *apiextv1.CustomResourceDefinition, reconciling bool) (*apiextv1.CustomResourceDefinition, bool, error) {
```
Comment (wking) [inline lib/resourceapply/apiext.go:16]: Can we pass in the mode instead of passing in a `reconciling` bool?  Like we do for ClusterOperators [here][1]?  That way we don't have to bump cluster signatures if we need to make some other mode distinction later on.

[1]: https://github.com/openshift/cluster-version-operator/blob/3b1047de4d2a82d2a1f4f92c224a1c46eba72ca9/pkg/cvo/internal/operatorstatus.go#L123-L126

File: lib/resourceapply/cv.go [hunk 2]
```diff
@@ -62,7 +36,7 @@ func ApplyClusterVersionFromCache(ctx context.Context, lister configlistersv1.Cl
 		return existing, false, nil
 	}
 
-	klog.V(2).Infof("Updating ClusterVersion %s due to diff: %v", required.Name, diff.ObjectDiff(existing, required))
+	klog.V(2).Infof("Updating ClusterVersion %s due to diff: %v", required.Name, cmp.Diff(existing, required))
```
Comment (wking) [inline lib/resourceapply/cv.go:39]: no mode guard here?

File: lib/resourceapply/imagestream.go [hunk 3]
```diff
@@ -11,7 +11,7 @@ import (
 	"k8s.io/utils/pointer"
 )
 
-func ApplyImageStreamv1(ctx context.Context, client imageclientv1.ImageStreamsGetter, required *imagev1.ImageStream) (*imagev1.ImageStream, bool, error) {
+func ApplyImageStreamv1(ctx context.Context, client imageclientv1.ImageStreamsGetter, required *imagev1.ImageStream, reconciling bool) (*imagev1.ImageStream, bool, error) {
 	existing, err := client.ImageStreams(required.Namespace).Get(ctx, required.Name, metav1.GetOptions{})
 	if errors.IsNotFound(err) {
 		actual, err := client.ImageStreams(required.Namespace).Create(ctx, required, metav1.CreateOptions{})
```
Comment (wking) [inline lib/resourceapply/imagestream.go:17]: no diff log before updating?

File: lib/resourceapply/apiext.go [hunk 4]
```diff
@@ -34,7 +34,9 @@ func ApplyCustomResourceDefinitionv1(ctx context.Context, client apiextclientv1.
 		return existing, false, nil
 	}
 
-	klog.V(2).Infof("Updating CRD %s due to diff: %v", required.Name, diff.ObjectDiff(existing, required))
+	if reconciling {
+		klog.V(2).Infof("Updating CRD %s due to diff: %v", required.Name, cmp.Diff(existing, required))
```
Comment (vrutkovs) [inline lib/resourceapply/apiext.go:38]: @LalatenduMohanty argued that we should use `v(4)` to avoid logging too many lines on production clusters (during development we could manually bump level and check for hotloops, as it doesn't need to be done too often)

File: lib/resourceapply/apiext.go (hunk 1 above)
Comment (arjunrn) [reply lib/resourceapply/apiext.go:16]: I considered that and decided against it because it would leak abstractions. `Mode` is part of the `resourcebuilder` package, where as the resource creation/update logic is in the `internal`. If in the future the full mode is required more refactoring can be done where the mode is moved to a common package and then it can be passed to the apply function.

File: lib/resourceapply/cv.go (hunk 2 above)
Comment (arjunrn) [reply lib/resourceapply/cv.go:39]: Unlike the other resources, the CV is created/updated from multiple controllers and the mode is not present there. So it cannot be used.

File: lib/resourceapply/apiext.go (hunk 4 above)
Comment (arjunrn) [reply lib/resourceapply/apiext.go:38]: I could reduce the log level but the logs will now be emitted only when there are resource updates when in reconcile mode. And this would be indicative of a hot loop. So these logs could help with diagnostics in production clusters. Changing the log level to 4 would be mean that before every release someone would have to increase the log level and check for hot loops. This would also not help in clusters where there are hot loops due to configuration which is not tested in CI.
In summary, this change basically eliminates all the logs which is considered verbose in the bug report. While at the same time keeping the logs for debugging in production clusters.

File: lib/resourceapply/imagestream.go (hunk 3 above)
Comment (arjunrn) [reply lib/resourceapply/imagestream.go:19]: Added now.

File: lib/resourceapply/apiext.go (hunk 4 above)
Comment (vrutkovs) [reply lib/resourceapply/apiext.go:38]: I'm fine with moving this to a different PR.

Lets file a bug for this change so that we could backport it? Other than that LGTM

**Review by vrutkovs — APPROVED** [review]
/lgtm

File: lib/resourceapply/apiext.go (hunk 4 above)
Comment (LalatenduMohanty) [reply lib/resourceapply/apiext.go:38]: > This would also not help in clusters where there are hot loops due to configuration which is not tested in CI.

This is not true as we are only concerned about hotloops because of manifests in release payload which CVO is responsible for. We can not track every hotloops in the cluster.

Comment (LalatenduMohanty) [reply lib/resourceapply/apiext.go:38]: /hold

> klog.V(2).Infof("Updating CRD %s due to diff: %v", required.Name, cmp.Diff(existing, required))

This will only catch hotloops from manifests which is managed by CVO. So it wont help if there is a hot loop somewhere else. Even if there is a hot loop in a production cluster it does not mean that the cluster's availability is impacted or this is a serious issue which needs to be fixed ASAP. Also I do not see much use for customers from this information because they can not change the manifests from release payload.

Comment (LalatenduMohanty) [reply lib/resourceapply/apiext.go:38]: However if we move the log level to V(4) then we can easily get it in our CI (by bumping the log level). So IMHO it should be atleast V(4).

Comment (arjunrn) [reply lib/resourceapply/apiext.go]: Updated to `V(4)`

**Review by LalatenduMohanty — APPROVED** [review]
/lgtm
